Quadratic elements for peripheral ring

miaoyinb opened the pull request:
closes #28169

miaoyinb committed:
Quadratic elements for peripheral ring

closes #28169

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2336597/) on 7ba0fe6 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28171/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28171/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format dc50da022633204fc54194633bd9bddc79c5644f`

miaoyinb committed:
civet patch

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2336691/) on 218c451 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/28171/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2336737/) on 218c451 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>dc50da</th>
      <th colspan=4>#28171 218c45</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.07%</td>
      <td>85.07%</td>
      <td>-0.00%</td>
      <td>80.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>104586</td>
      <td>104602</td>
      <td>+16</td>
      <td>36</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18359</td>
      <td>18365</td>
      <td>+6</td>
      <td>9</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28171/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28171/coverage/framework)


## Modules coverage

### Reactor

<table>
  <thead>
    <tr>
      <th></th>
      <th>dc50da</th>
      <th colspan=4>#28171 218c45</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>93.68%</td>
      <td>93.68%</td>
      <td>+0.01%</td>
      <td>95.51%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>7272</td>
      <td>7308</td>
      <td>+36</td>
      <td>85</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>491</td>
      <td>493</td>
      <td>+2</td>
      <td>4</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28171/coverage_diff/reactor)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28171/coverage/reactor)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/28171/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/28171/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/28171/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/28171/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/28171/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/28171/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/28171/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/28171/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/28171/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/28171/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/28171/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/28171/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/28171/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/28171/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/28171/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/28171/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/28171/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/28171/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/28171/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/28171/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/28171/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/28171/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/28171/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/28171/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/28171/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/28171/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/28171/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/28171/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 80.00% is less than the suggested 90.0%



This comment will be updated on new commits.

miaoyinb commented:
@GiudGiud FYI. Currently it can only do QUAD4 and QUAD9, I can add QUAD8 if needed.

miaoyinb commented:
@eshemon FYI

GiudGiud commented on modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/tests:
```diff
@@ -77,6 +77,15 @@
+    input = 'ring_quadratic.i'
+    exodiff = 'ring_boundary_layers_quadratic_in.e'
+    cli_args = '--mesh-only ring_boundary_layers_quadratic_in.e'
+    requirement ='The system shall be able to add a peripheral ring consistin of quandratic with boundary layers to the input mesh and preserve volumes even with polygonization.'
```
```suggestion
    requirement ='The system shall be able to add a peripheral ring consisting of quadratic elements with boundary layers to the input mesh and preserve volumes despite polygonization.'
```

GiudGiud commented on framework/include/utils/FillBetweenPointVectorsTools.h:
```diff
@@ -296,6 +296,31 @@ bool isCurveOpenSingleSegment(MeshBase & mesh,
  * @param mesh input mesh that contains the node series to be examined
  * @param max_node_radius the maximum radius of the nodes in the series
  * @param ordered_node_list the ordered node ids on the given curve
+ * @param node_assm the assembly of the node id pairs that correspond vertices of a series of sides
```
```suggestion
 * @param node_assm the assembly of the node id pairs that correspond to vertices of the sides on the closed loop
```

GiudGiud replied:
if it s correct

GiudGiud commented on framework/include/utils/FillBetweenPointVectorsTools.h:
```diff
@@ -296,6 +296,31 @@ bool isCurveOpenSingleSegment(MeshBase & mesh,
+ * @param midpoint_node_list the node ids of the midpoints of the sides for quadratic sides
+ * @param origin_pt origin position of the given mesh (used for azimuthal angle calculation)
+ * @param is_closed_loop whether the series of nodes form a closed loop with consecutive nodes's
+ * azimuthal angles change monotonically
```
```suggestion
 * azimuthal angles changing monotonically
```

GiudGiud reviewed (commented):
will look at the code latr

miaoyinb committed:
Apply suggestions from code review

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2338469/) on a771233 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28171/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28171/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format dc50da022633204fc54194633bd9bddc79c5644f`

miaoyinb committed:
civet patch

moosebuild commented:
Job [Documentation public apps](https://civet.inl.gov/job/2338534/) on 2eeca9c : invalidated by @miaoyinb

pbehne committed:
Defined and used some intermediate variables to improve readability. Ref #28169

pbehne committed:
Changed loops to range-based per MOOSE software standards. Ref #28169

pbehne commented on modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:
```diff
@@ -150,46 +150,88 @@ PeripheralRingMeshGenerator::generate()
+      else if (etype == EDGE3)
+        quadratic_side = true;
+      else
+        paramError("input", "Input contains elements that are not supported by this generator.");
```
Please add a `RunException` test for this.

miaoyinb replied:
Does current MOOSE have an easy way to load/create a third-order 2D mesh to trigger this error?

pbehne replied:
Tagging @roystgnr on this one since I'm not sure. He will likely know off the top of his head.

pbehne replied:
I see that libmesh has a 1D `Edge4` element. I'll see if I can figure out what kind of element uses this.

pbehne replied:
As far as I can tell from looking through libmesh source, `Edge4` is not used as a side in any element. I'll let Roy have the final word, though.

miaoyinb replied:
Right, I do not see any complete third-order 2D elements there.

roystgnr replied:
We currently don't have any 3rd-order elements outside of 1D, so no way to add test coverage for this.

We might be adding some in the next year, so it's good that we're catching that possibility here, but not soon enough that I expect I'll remember this PR when we're finally able to add a test for it.

pbehne commented on modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:
```diff
@@ -150,46 +150,88 @@ PeripheralRingMeshGenerator::generate()
+    }
+  }
+  if (linear_side && quadratic_side)
+    paramError("input", "Input contains mixed element types on the external boundary.");
```
Please add a `RunException` test for this.

pbehne commented on modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:
```diff
@@ -440,14 +503,32 @@ PeripheralRingMeshGenerator::innerBdryLayerNodesDefiner(
     // neighboring node-to-origin vectors. Flipped nodes are marked and excluded during the
     // follow-up interpolation.
-    if (!MooseUtils::absoluteFuzzyEqual(input_bdry_angles[i], M_PI / 2.0))
+    if (!MooseUtils::absoluteFuzzyEqual(input_bdry_angles[i], M_PI / 2.0, 1.0e-4))
```
Why the tolerance change?

miaoyinb replied:
The previous `PeripheralRingMeshGenerator::innerBdryLayerNodesDefiner()` has a bug. It would have issue in some cases when the inner boundary layer is too thick. This, along with the following additions, is part of the fix.

pbehne commented on modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/ring_quadratic.i:
```diff
@@ -0,0 +1,33 @@
+      preserve_volumes = on
+
+      tri_element_type = TRI6
+      quad_element_type = QUAD8
```
Why `QUAD8` instead of `QUAD9`? Just curious.

GiudGiud replied:
this PR adds quad9 for the periphery. for the mesh inside, quad8 will use less memory and is just as compatible with the boundary

pbehne reviewed (changes requested):
Please put a short description of what this PR does in the header of the PR, even if it is just copying and pasting the issue.

While reviewing, I cleaned up some of the changed files (not really related to your PR). I'd like to get these changes in with this PR. May I push to your branch (you may have to add me as a collaborator to your fork of MOOSE). If not, I can also make a PR with my changes to your fork (after you add me as a collaborator).

miaoyinb commented:
> Please put a short description of what this PR does in the header of the PR, even if it is just copying and pasting the issue.
> 
> While reviewing, I cleaned up some of the changed files (not really related to your PR). I'd like to get these changes in with this PR. May I push to your branch (you may have to add me as a collaborator to your fork of MOOSE). If not, I can also make a PR with my changes to your fork (after you add me as a collaborator).

I think the PR template says it does not need any description if the issue # is linked in the header. Anyway I will add you as a collaborator, please do a PR there so that I could take a look. Thanks.

miaoyinb committed:
add a test to check mixed side element types

pbehne commented:
> > Please put a short description of what this PR does in the header of the PR, even if it is just copying and pasting the issue.
> > While reviewing, I cleaned up some of the changed files (not really related to your PR). I'd like to get these changes in with this PR. May I push to your branch (you may have to add me as a collaborator to your fork of MOOSE). If not, I can also make a PR with my changes to your fork (after you add me as a collaborator).
> 
> I think the PR template says it does not need any description if the issue # is linked in the header. Anyway I will add you as a collaborator, please do a PR there so that I could take a look. Thanks.

Thanks for pointing out the template instructions. Since you have an issue number, could you go ahead and delete the empty headers (Reason, Design, Impact)?

PR is up here: https://github.com/miaoyinb/moose/pull/5

pbehne reviewed (approved):
Thanks for adding this feature! I'm ready to merge if @GiudGiud is.

miaoyinb commented:
@GiudGiud Thank you for reviewing/merging my other PR. Just a friendly reminder that this PR is still pending.

GiudGiud commented:
sorry for the delay. I am setting up a new machine the other one died by IT

GiudGiud commented on modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/ring_quadratic.i:
```diff
@@ -0,0 +1,33 @@
+
+      preserve_volumes = on
+
+      tri_element_type = TRI6
```
If I  comment these two out to make the mesh linear, I get a different volumes for the external ring

miaoyinb replied:
I will look into this, thanks

miaoyinb replied:
Interesting, paraview says the areas are the same...(706.858 total which is the same as the linear elements values reported by MOOSE)

Maybe we need to take a look at the source code that calculates these printed volume values...

miaoyinb replied:
```
[Mesh]
  [dir1]
    type = ElementGenerator
    nodal_positions = '0 0 0
                       1 0 0
                       1 1 0
                       0 1 0
                       0.5 -0.5 0
                       1 0.5 0
                       0.5 1 0
                       0 0.5 0
                       0.5 0.5 0'
    element_connectivity = '0 1 2 3 4 5 6 7 8'
    elem_type = 'QUAD9'
  []
[]
```
The element area is 1.25 but MOOSE says 1.333 (or quadratic element area has a different definition?)

GiudGiud replied:
@roystgnr thoughts?

GiudGiud replied:
I think when you say 1.25 you have a triangle between the three slopes at the bottom
shouldnt it be quadratic?

miaoyinb replied:
I just tried to give an extreme case. There might be some rules that I am not aware of. Let's see if @roystgnr has any insights.

miaoyinb replied:
> I think when you say 1.25 you have a triangle between the three nodes at the bottom of the element shouldnt it be quadratic?

Ah, so the edge should be parabolic for QUAD9 element. In that case the radius correction function needs to be modified.

GiudGiud replied:
I think so.
Irrelevant here but I think it might be the same for quad8?

miaoyinb replied:
fixed. sorry for the batch regolding...

roystgnr replied:
> Ah, so the edge should be parabolic for QUAD9 element

This is correct.  In the `ElementGenerator` case in particular, 4/3 is the correct area.

GiudGiud reviewed (commented):
there's something I dont understand with the test

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2379866/) on b6a5a3b wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28171/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28171/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format e0de6598545e9ba8ea902d1ef1ef2560b05decc2`

GiudGiud commented on modules/reactor/test/tests/meshgenerators/patterned_hex_peripheral_modifier/gold/single_hex_quadratic_in.e:
the nodes should not have moved for those? the modifying of the periphery was sqauree right?

GiudGiud replied:
true true

GiudGiud commented on modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/gold/ring_boundary_layers_quadratic_in.e:
let's remove that gold file from commit #1 to avoid the regold

GiudGiud commented on modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C:
```diff
@@ -1138,23 +1138,68 @@ PolygonMeshGeneratorBase::quadElemDef(ReplicatedMesh & mesh,
+    }
+    return std::sqrt(tmp_acc_azi / tmp_acc);
+  }
+  else // order == 2
```
@roystgnr for a quad9, does the position of a middle node influence the shape of the outer boundary?

roystgnr replied:
It does not.  Not for Lagrange mappings, not for Rational Bezier Bernstein mappings.

There are mappings with less locality than the ones we use, but we'll probably never support any directly, only indirectly the way we support IsoGeometric Analysis NURBS mappings - via extraction operators that give more locally-defined elements to integrate on.

GiudGiud replied:
thanks

how exact is the volume computation for second order elements in libmesh? I still see tiny diffs (1e-5) with the correction we have here

miaoyinb replied:
I think that summation of small $\sin^3(\theta)$ compromises the errors...

GiudGiud replied:
could be

are you summing two separate quadratic bands, separated at the mid point of the edge?
I think the whole edge is covered by a single quadratic band?

GiudGiud replied:
> But are you diffing against the exact biquadratic volume, or against a perfect circle? Biquadratics (in Lagrange; I'm working on RBB mesh generator tools) aren't enough to exactly mesh a circle; no polynomial mapping is. The difference between a unit circle and its biquadratic approximation is probably more like 1e-5.

I am diffing against the corrected 1st order mesh, which should match the perfect circle volume exactly already.
While the bi-quadratics do not exactly mesh the circle, we are correcting them here so that the volume is the exact same. This is what this routine is intended for. So yes it's only correcting a very small error already

miaoyinb replied:
> could be
> 
> are you summing two separate quadratic bands, separated at the mid point of the edge? I think the whole edge is covered by a single quadratic band?

Yes, it's a single band determined by the three nodes.

miaoyinb replied:
If we really want to eliminate that 1e-6 relative error, one possible way would be to make a dummy TRI6 mesh with the same external nodes and use its area as the reference to derive a correction factor.

GiudGiud replied:
It s more that if we expect the error to be 1e-6 already with quad-elements (as Roy mentioned), then why do we still see that after correction

I m yet to run this again and make up my mind on this

GiudGiud replied:
it might just be that the error is a larger than expected pre-correction

miaoyinb replied:
I will also run more cases to get a clearer idea what is going on.

GiudGiud commented on modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C:
```diff
@@ -1138,23 +1138,68 @@ PolygonMeshGeneratorBase::quadElemDef(ReplicatedMesh & mesh,
+    {
+      tmp_acc += std::sin(azi_interval_list[i].first + azi_interval_list[i].second);
+      tmp_acc +=
+          4.0 / 3.0 * std::sin((azi_interval_list[i].first + azi_interval_list[i].second) / 2) *
+          std::sin((azi_interval_list[i].first + azi_interval_list[i].second) / 2) *
+          std::sin((azi_interval_list[i].first + azi_interval_list[i].second) / 2) /
```
MathUtils
```suggestion
          4.0 / 3.0 * Utility::pow<3>(std::sin((azi_interval_list[i].first + azi_interval_list[i].second) / 2)) /
```

GiudGiud commented on modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C:
```diff
@@ -1138,23 +1138,68 @@ PolygonMeshGeneratorBase::quadElemDef(ReplicatedMesh & mesh,
+          std::sin((azi_interval_list[i].first + azi_interval_list[i].second) / 2) *
+          std::sin((azi_interval_list[i].first + azi_interval_list[i].second) / 2) /
+          std::cos(azi_interval_list[i].first / 2) / std::cos(azi_interval_list[i].second / 2);
+      tmp_acc_azi += azi_interval_list[i].first + azi_interval_list[i].second;
```
why both for this case?

pbehne commented on framework/src/utils/FillBetweenPointVectorsTools.C:
```diff
@@ -466,7 +467,7 @@ surrogateGenerator(std::vector<Real> & weighted_surrogate_index,
       weighted_surrogate_index[j] +=
           (unweighted_surrogate_index[j] - *(it_1 - 1)) * wt[it_start + it_dist - 1];
-      for (unsigned int k = 1; k < it_dist; k++)
+      for (const auto k : make_range(long(1), it_dist))
```
Here is the compiler error when I don't cast to long:
`In file included from /Users/behnpa/projects/moose_quadratic_ring/framework/build/unity_src/utils_Unity.C:20:
/Users/behnpa/projects/moose_quadratic_ring/framework/src/utils/FillBetweenPointVectorsTools.C:470:27: error: no matching function for call to 'make_range'
      for (const auto k : make_range(1, it_dist))`
I tried using auto instead of long, and got:
`/Users/behnpa/projects/moose_quadratic_ring/framework/src/utils/FillBetweenPointVectorsTools.C:470:38: error: 'auto' not allowed here
      for (const auto k : make_range(auto(1), it_dist))
`

pbehne replied:
For loops that start from 1, should we just use the other style of for loop?

GiudGiud replied:
unless you find a way to get the right data type for the 1 with various compiling options I would say yes

miaoyinb replied:
could be overkill but this would work...

`make_range(castIndexForRange(foo,1),foo)`

```
template <typename T, typename P>
const T
castIndexForRange(const T &, //type_input
                  const P & value_input)
{
  return (T)(value_input);
}
```

roystgnr replied:
`make_range()` is just some syntactic sugar.  It's just to make it easier to avoid the typical C-style mistake of `for (foo i=begin(); i != expensive_call_that_shouldnt_be_here(); ++i)`, not because there's anything else we could fix with range-for vs C-style-for loops.  (though I'll bet we have a lot of both kinds of loops that should be `std::foreach`, and I'm thinking of wrapping that to play with different `std::execution::` options...)

`castIndexForRange()` is syntactic Carolina Reaper powder.  Maybe it's an acquired taste?

I think C-style is fine; otherwise `1l` is more terse than `long(1)`

GiudGiud commented:
saw your changes, I ll take a look tomorrow

GiudGiud commented:
It looks like the background size in the square lattice, when deformation of non-circular regions is set to false, is different between quadratic and linear elements. It affects the regolded files for patterned_cartesian_mg (pattern_sq_) and I think that's it.

Which volume is most important to preserve in that case?

GiudGiud commented:
Otherwise we are good on all fronts here

miaoyinb commented:
> It looks like the background size in the square lattice, when deformation of non-circular regions is set to false, is different between quadratic and linear elements. It affects the regolded files for patterned_cartesian_mg (pattern_sq_) and I think that's it.
> 
> Which volume is most important to preserve in that case?

If the non-circular region deformation is activated, the background region is deformed to allow small hexagon/square size that cuts the original patterned unit cells. In that case, we do not care about the volume preserve as they are deformed anyway.

But if the deformation is not activated, the midpoint should be colinear with the vertices and the quandratic elements and linear elements should have the same sizes(?)

GiudGiud commented:
I think the square size should be the same for both linear and quadractic though? We lose the volume conservation of the background but we impose a square size right?

the diff is in the case with the deformation

miaoyinb commented:
> I think the square size should be the same for both linear and quadractic though? We lose the volume conservation of the background but we impose a square size right?
> 
> the diff is in the case with the deformation

Right.

GiudGiud commented:
ok so it's not tied to this PR but seems there's a problem in the PatternedCartesian w/ deformation & second order.
pls let me know if you want to address it here with the regold or if we should create an issue

miaoyinb commented:
Ah, I figured out the source of the issue. Let's address it here.

miaoyinb committed:
correct the quadratic radius preseve factor and regolding
correct the quadratic radius preseve factor and regolding

code cleanup and more comments to address the review comments

use c-style loop for nontrivial cases

use libmesh tri6 volume for radius correction

fix patterned peripheral modifier

miaoyinb commented:
That should be fixed now.

GiudGiud commented:
this didnt fix it. If you zoom in on the corner you can see that pattern_sq.e and pattern_sq_quad8.e for the patterned_cartesian_mg dont line up at the pin background / duct background interface

miaoyinb commented:
> this didnt fix it. If you zoom in on the corner you can see that pattern_sq.e and pattern_sq_quad8.e for the patterned_cartesian_mg dont line up at the pin background / duct background interface

Let me check it again. It seems the hex and cartesian have different issues...

miaoyinb commented:
The boundary of the pin background and assembly background after deformation is okay to be different between linear and quadratic elements. The size of the outer boundary of the assembly (i.e., inner size of the innermost duct) is important and must be consistent.

GiudGiud reviewed (approved):
thanks for the contribution!
